File: Recover-Binary-Search-Tree/Recover-Binary-Search-Tree.py


# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a tree node
    def recoverTree(self, root):
        self.pre = None
        self.first = None
        self.second = None
        self.inorder(root)
        if self.first!=None and self.second!=None:
            self.first.val, self.second.val = self.second.val, self.first.val
        return root
    def inorder(self,root):
        if root==None:
            return
        
        self.inorder(root.left)
        if self.pre!=None and self.pre.val>root.val:
            if self.first==None:
                self.first = self.pre
                self.second = root
            else:
                self.second = root
        self.pre = root
        self.inorder(root.right)
 

File: Recover-Binary-Search-Tree/Recover-Binary-Search-Tree.cpp

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *pre;
    TreeNode *first;
    TreeNode *second;
    void inorder(TreeNode *root)
    {
        if(!root)
            return;
        inorder(root->left);
        if(pre && pre->val>root->val)
        {
            if(!first)
            {
                first=pre;
                second=root;
            }
            else
                second=root;
        }
        pre=root;
        inorder(root->right);
    }
    void recoverTree(TreeNode *root) {
        inorder(root);
        if(first)
            swap(first->val,second->val);
    }
};


 